lib/modules/manager/pipenv/artifacts.ts

```diff
@@ -113,35 +114,100 @@ export function getPipenvConstraint(
   return '';
 }
 
-function getMatchingHostRule(url: string): HostRule {
-  return find({ hostType: PypiDatasource.id, url });
+export function getMatchingHostRule(url: string): HostRule | null {
+  const parsedUrl = parseUrl(url);
+  if (parsedUrl) {
+    parsedUrl.username = '';
+    parsedUrl.password = '';
+    const urlWithoutCredentials = parsedUrl.toString();
+
+    return find({ hostType: PypiDatasource.id, url: urlWithoutCredentials });
+  }
+  return null;
 }
 
-async function findPipfileSourceUrlWithCredentials(
+async function findPipfileSourceUrlsWithCredentials(
   pipfileContent: string,
   pipfileName: string,
-): Promise<string | null> {
+): Promise<URL[]> {
   const pipfile = await extractPackageFile(pipfileContent, pipfileName);
-  if (!pipfile) {
-    logger.debug('Error parsing Pipfile');
-    return null;
-  }
 
-  const credentialTokens = [
-    '$USERNAME:',
-    // eslint-disable-next-line no-template-curly-in-string
-    '${USERNAME}',
-    '$PASSWORD@',
-    // eslint-disable-next-line no-template-curly-in-string
-    '${PASSWORD}',
-  ];
+  return (pipfile?.registryUrls
+    ?.map(parseUrl)
+    ?.filter((url) => !!url?.username) ?? []) as URL[]; //tsc does not notice null entries are removed
+}
+
+/**
+ * This will extract the actual variable name from an environment-placeholder:
+ * ${USERNAME:-defaultvalue} will yield 'USERNAME'
+ */
+export function extractEnvironmentVariableName(
+  credential: string,
+): string | null {
+  const match = decodeURI(credential).match('([a-zA-Z0-9_]+)');
+  return match?.length ? match[0] : null;
+}
 
-  const sourceWithCredentials = pipfile.registryUrls?.find((url) =>
-    credentialTokens.some((token) => url.includes(token)),
+export function addExtraEnvVariable(
+  extraEnv: ExtraEnv<unknown>,
+  environmentVariableName: string,
+  environmentValue: string,
+): void {
+  logger.debug(
```

Comment:
```suggestion
  logger.trace(
```

you can remap via config if required

https://docs.renovatebot.com/configuration-options/#loglevelremap